DOC-965 Deprecate and refactor redpanda_migrator fields (#159)
redpanda-data/rp-connect-docs: modules/components/pages/outputs/redpanda_migrator.adoc

  label: ""
  redpanda_migrator:
    seed_brokers: [] # No default (required)
    client_id: benthos
    tls:
      enabled: false
      skip_cert_verify: false
      enable_renegotiation: false
      root_cas: ""
      root_cas_file: ""
      client_certs: []
    sasl: [] # No default (optional)
    topic: "" # No default (required)
    key: "" # No default (optional)
    partitioner: "" # No default (optional)
    partition: ${! meta("partition") } # No default (optional)
    metadata:
      include_prefixes: []
      include_patterns: []
    metadata_max_age: 5m
    timestamp_ms: ${! timestamp_unix_milli() } # No default (optional)
    max_in_flight: 256
    input_resource: redpanda_migrator_input
    replication_factor_override: true
    replication_factor: 3
    translate_schema_ids: true
    schema_registry_output_resource: schema_registry_output
    idempotent_write: true
    compression: "" # No default (optional)  
    timeout: 10s
    max_message_bytes: 1MB
    broker_write_max_bytes: 100MB
```

--
======

> You must specify the label of the associated xref:components:inputs/redpanda_migrator.adoc[`redpanda_migrator` input] in the <<input_resource,`input_resource` field>>. When connected, this output queries the `redpanda_migrator` input for topic and ACL configurations and attempts to create them on the specified Kafka broker. If the broker does not contain the current message topic, this output attempts to create the topic along with its ACLs.

Reviewer: > and attempts to create them on the specified Kafka broker.
Guess that's a bit redundant

A plain text password for when the private key is password encrypted in PKCS#1 or PKCS#8 format. The obsolete `pbeWithMD5AndDES-CBC` algorithm is not supported for the PKCS#8 format.

> WARNING: The `pbeWithMD5AndDES-CBC` algorithm does not authenticate ciphertext, and is vulnerable to padding Oracle attacks which may allow an attacker recover the plain text password.

Reviewer: Nit: This is not related to Oracle the company. Details here: https://en.wikipedia.org/wiki/Padding_oracle_attack

> WARNING: The `pbeWithMD5AndDES-CBC` algorithm does not authenticate ciphertext, and is vulnerable to padding oracle attacks which may allow an attacker recover the plain text password.

Reviewer 2: ```suggestion
WARNING: The `pbeWithMD5AndDES-CBC` algorithm does not authenticate ciphertext, and is vulnerable to padding oracle attacks which may allow an attacker to recover the plain text password.
```

Author: Updated.

An external ID to provide when assuming a role.

*Type*: `string`

*Default*: `""`

=== `topic`

A topic to write messages to.

This field supports xref:configuration:interpolation.adoc#bloblang-queries[interpolation functions].

*Type*: `string`

=== `key`

An optional key to populate for each message.

This field supports xref:configuration:interpolation.adoc#bloblang-queries[interpolation functions].

*Type*: `string`

=== `partitioner`

Override the default murmur2 hashing partitioner.

*Type*: `string`

|===
| Option | Summary

| `least_backup`
| Chooses the least backed up partition (the partition with the fewest buffered records). Partitions are selected per batch.
| `manual`
| Manually select a partition for each message. You must also specify a value for the `partition` field.
| `murmur2_hash`
| Kafka's default hash algorithm that uses a 32-bit murmur2 hash of the key to compute the partition for the record.
| `round_robin`

> | Round-robin's messages through all available partitions. This algorithm has lower throughput and causes higher CPU load on brokers, but is useful if you want to ensure an even distribution of records to partitions.

Reviewer 2: As written, this is possessive, which makes no sense. I think you are intending to use **round-robin** as a verb? Perhaps "Use round-robin messages through..."?

Author: Agreed. Updated

|===

=== `partition`

An optional explicit partition to set for each message. This field is only relevant when the `partitioner` is set to `manual`. The provided interpolation string must be a valid integer.

This field supports xref:configuration:interpolation.adoc#bloblang-queries[interpolation functions].

*Type*: `string`

```yml
# Examples

partition: ${! meta("partition") }
```

=== `metadata`

Specify which (if any) metadata values are added to messages as headers.

*Type*: `object`

=== `metadata.include_prefixes`

Provide a list of explicit metadata key prefixes to match against.

*Type*: `array`

*Default*: `[]`

```yml
# Examples

include_prefixes:
  - foo_
  - bar_

include_prefixes:
  - kafka_

include_prefixes:
  - content-
```

=== `metadata.include_patterns`

Provide a list of explicit metadata key regular expression (re2) patterns to match against.

*Type*: `array`

*Default*: `[]`

```yml
# Examples

include_patterns:
  - .*

include_patterns:
  - _timestamp_unix$
```

=== `metadata_max_age`

> The maximum period of time after which metadata is refreshed. Reducing this value increases the frequency with which newly created topics are identified. 

Reviewer 2: ```suggestion
The maximum period of time after which metadata is refreshed. Reducing this value increases the frequency with which newly-created topics are identified. 
```

Author: Updated

timestamp_ms: ${! metadata("kafka_timestamp_ms") }
```

=== `max_in_flight`

The maximum number of batches to send in parallel at any given time.

*Type*: `int`

*Default*: `256`

=== `input_resource`

The label of the xref:components:inputs/redpanda_migrator.adoc[`redpanda_migrator` input] from which to read the configurations of topics and ACLs for creation.

*Type*: `string`

*Default*: `redpanda_migrator_input`

=== `replication_factor_override`

Whether to override the replication factor of input topics when creating copies of them on the output cluster. You must specify the new replication factor in the <<replication-factor, `replication_factor` field>>.

*Type*: `bool`

*Default*: `true`

[[replication-factor]]
=== `replication_factor`

The replication factor for created topics. This is only used when `replication_factor_override` is set to `true`.

*Type*: `int`

*Default*: `3`

=== `translate_schema_ids`

> When set to `true`, this field translates the schema ID in each message ready to write to the destination schema registry.

Reviewer: `this field` -> `this output`

Maybe we want to say something about how this translation works, dunno. I'm thinking to leave that for a future discussion, but this output stores a mapping of source -> destination schema registry IDs in memory and uses that mapping to perform the translation.

Author: OK. I'm going to create a ticket for this now for the next release. [DOC-977](https://redpandadata.atlassian.net/browse/DOC-977).

> When set to `true`, this field translates the schema ID in each message ready to write to the destination schema registry.

Reviewer 2: Can't make a good suggestion really, because it is just not clear.

> When set to `true`, this field translates the schema ID in each message ready to write to the destination schema registry.

Author: This is going to be resolved as part of another ticket. It does need better explanation: https://redpandadata.atlassian.net/browse/DOC-977

*Type*: `bool`

*Default*: `true`

=== `schema_registry_output_resource`

The label of the xref:components:outputs/schema_registry.adoc[`schema_registry` output] to use for fetching schema IDs.

*Type*: `string`

*Default*: `schema_registry_output`

=== `rack_id`

A rack identifier for this client.

*Type*: `string`

*Default*: `""`

=== `idempotent_write`

> Enable the idempotent write producer option. This requires the `IDEMPOTENT_WRITE` permission on `CLUSTER`. Disable this option if the `IDEMPOTENT_WRITE` permission is not available.

Reviewer 2: ```suggestion
Enable the idempotent write producer option. This requires the `IDEMPOTENT_WRITE` permission on `CLUSTER`. Disable this option if the `IDEMPOTENT_WRITE` permission is unavailable.
```

Author: Updated.

*Type*: `bool`

*Default*: `true`

=== `compression`

Set an explicit compression type (optional). The default preference is to use `snappy` when the broker supports it. Otherwise, use `none`.

*Type*: `string`

Options:
`lz4`
, `snappy`
, `gzip`
, `none`
, `zstd`

=== `timeout`

The maximum period of time to wait for message sends before abandoning the request and retrying.

*Type*: `string`

*Default*: `10s`

=== `max_message_bytes`

The maximum space in bytes that an individual message may use. Messages larger than this value are rejected. This field corresponds to Kafka's `max.message.bytes`.

*Type*: `string`

*Default*: `"1MB"`

```yml
# Examples

max_message_bytes: 100MB

max_message_bytes: 50mib
```

=== `broker_write_max_bytes`

The upper bound for the number of bytes written to a broker connection in a single write. This field corresponds to Kafka's `socket.request.max.bytes`.

*Type*: `string`

*Default*: `100MB`

```yml
# Examples

broker_write_max_bytes: 128MB

broker_write_max_bytes: 50mib

> ```

Reviewer: Unfortunately these will change again for all franz-go-based outputs since https://github.com/redpanda-data/connect/pull/3150 was merged. Guess it's OK to tackle that in another round.

Author: Yes, all changes here are for 4.45.0, so we'll need to update again.
